lib/Authentication.ts

```diff
@@ -49,6 +52,150 @@ export default class Authentication {
+    let key: PrivateKey | undefined;
+    key = this.getKey(requesterDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${requesterDid}`);
+    }
+
+    const token = new JwsToken(request, this.factory);
+    return token.sign(key);
+  }
+
+  /**
+   * Verifies signature on request and returns AuthenticationRequest.
+   * @param request Authentiation Request as a buffer or string.
+   */
+  public async verifyAuthenticationRequest (request: Buffer | string): Promise<AuthenticationRequest> {
+    // identity function in place of later signature verification
+    let jwsToken: JwsToken;
+    if (request instanceof Buffer) {
+      jwsToken = new JwsToken(request.toString(), this.factory);
+    } else {
+      jwsToken = new JwsToken(request, this.factory);
+    }
+    const keyId = jwsToken.getHeader().kid;
+    const keyDid = DidDocument.getDidFromKeyId(keyId);
+    const results = await this.resolver.resolve(keyDid);
+    const didPublicKey = results.didDocument.getPublicKey(keyId);
+    if (!didPublicKey) {
+      throw new Error('Could not find public key');
+    }
+    const publicKey = this.factory.constructPublicKey(didPublicKey);
+    const content = await jwsToken.verifySignature(publicKey);
+
+    const verifiedRequest: AuthenticationRequest = JSON.parse(content);
+    if (verifiedRequest.iss !== keyDid) {
+      throw new Error('Signing DID does not match issuer');
+    }
+    return verifiedRequest;
+  }
+
+  /**
+   * Given a challenge, forms a signed response using a given DID that expires at expiration, or a default expiration.
+   * @param authRequest Challenge to respond to
+   * @param responseDid The DID to respond with
+   * @param claims Claims that the requester asked for
+   * @param expiration optional expiration datetime of the response
+   */
+  public async formAuthenticationResponse (authRequest: AuthenticationRequest, responseDid: string, claims: any, expiration?: Date): Promise<string> {
+    let key: PrivateKey | undefined;
+    key = this.getKey(responseDid);
+    if (!key) {
+      throw new Error(`Could not find a key for ${responseDid}`);
+    }
+
+    const publicKey: PublicKey = key.getPublicKey();
+
+    // milliseconds to seconds
+    const milliseconds = 1000;
+    if (!expiration) {
+      const expirationTimeOffsetInMinutes = 5;
+      expiration = new Date(Date.now() + milliseconds * 60 * expirationTimeOffsetInMinutes); // 5 minutes from now
+    }
+    const iat = Math.floor(Date.now() / milliseconds); // ms to seconds
+    const response: AuthenticationResponse = {
+      iss: 'https://self-issued.me',
+      sub: responseDid,
+      aud: authRequest.client_id,
+      nonce: authRequest.nonce,
+      exp: Math.floor(expiration.getTime() / milliseconds),
+      iat,
+      sub_jwk: {},
+      did: responseDid,
+      state: authRequest.state
+    };
+
+    let authResponse = Object.assign(response, { sub_jwk: publicKey });
+    authResponse = Object.assign(authResponse, claims);

```

Comment:
You're assigning anything in claims as claims? So I could pass a claims object:
```
{
  iss: https://not-self-issued.me
}
```
override the issuer and thats expected behavior?

Reply:
Interesting, there should probably be a guard for JWT claims such as **_sub_** and **_iss_**